### PR TITLE
Task List Section Locking Subscription

### PR DESCRIPTION
**Pull request by OddTomBrooks**
# EASI-1995

## Changes and Description

- Added a PubSub service
- Added GQL subscriptions for task list section locking

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Tested user-facing changes with voice-over and the [rotor menu](https://support.apple.com/guide/voiceover/with-the-voiceover-rotor-mchlp2719/mac)

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

File: pkg/shared/pubsub/event.go [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package pubsub
+
+// Event provides the identifying information necessary for a subscription model
+type Event string
```
Comment (StevenWadeOddball) [inline pkg/shared/pubsub/event.go:4]: Maybe EventType would be a clearer name here?

File: pkg/graph/schema.resolvers.go [hunk 2]
```diff
@@ -237,6 +237,27 @@ func (r *mutationResolver) DeleteDiscussionReply(ctx context.Context, id uuid.UU
...
+
+	return true, nil
+}
+
+func (r *mutationResolver) UnlockTaskListSection(ctx context.Context, modelPlanID uuid.UUID, section string) (bool, error) {
+	principal := appcontext.Principal(ctx).ID()
+
+	resolvers.UnlockTaskListSection(r.pubsub, modelPlanID, section, principal)
+
+	return true, nil
```
Comment (StevenWadeOddball) [inline pkg/graph/schema.resolvers.go:253]: I think this is potentially very confusing to always return true. 

I locked the task as USER MINT, tried to unlock as user COOL. I could see that it didn't unlock, but I got a response that suggested that it did unlock the task list section. I think maybe there should be a different return type for this struct? This also is swallowing the error
image

I think the paradigm for the other resolvers is to have the resolver return the return object and error, instead of building it here in the schema resolvers file.

File: pkg/graph/schema.graphql [hunk 3]
```diff
@@ -606,6 +617,26 @@ updateDiscussionReply(id: UUID!, changes: DiscussionReplyChanges!): DiscussionRe
 
 deleteDiscussionReply(id: UUID!): DiscussionReply!
 @hasRole(role: MINT_BASE_USER)
+
+lockTaskListSection(modelPlanID: UUID!, section: String!): Boolean!
```
Comment (StevenWadeOddball) [inline pkg/graph/schema.graphql:621]: I think we should have a task list section field here. In your example, we could lock any string we wanted. 

I don't think there is any point to allow users to send a request for a section that doesn't exist.

File: pkg/graph/resolvers/plan_task_list_section_locks.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		return false
+	}
+
+	_, found = session[section]
+	return found
+}
+
+// GetTaskListSectionLocks is a convenience relay method to call the corresponding method on a resolver implementation
+func GetTaskListSectionLocks(modelPlanID uuid.UUID) []*model.TaskListSectionLockStatus {
+	return NewPlanTaskListSectionLocksResolverImplementation().GetTaskListSectionLocks(modelPlanID)
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/plan_task_list_section_locks.go:144]: Can you speak to why you chose to use the implementation method here? I feel like it might be clearer to just to duplicate making the object instance here. It's nice that this can be called in one line, but I think it would be clearer to me without the function call. (It feels a lot more complex)

**Review by StevenWadeOddball — CHANGES_REQUESTED** [review]
A lot of great stuff here! I think that this subscription system will open a lot of doors for us.  I still need to review more, but I have a little bit of feedback (see the comments below). I wasn't able to test the subscription portion yet, could you provide a bit more data about what the URL you send the web socket init and subscribe methods to?

I would also really like to see an ADR and some sort of documentation for this. From our conversations about subscriptions, I know you have put a lot of excellent thought into why you went with this technology, and considerations  we should have for the future. I think it is important to document it here while it is still fresh. 

Can you also speak to how you envision the task list locking to go? It seems that this really is just a method for the front end to call, but we don't enforce it anywhere on the backend.  I think it could be worth a bit of extra time on this portion. 

Another possible approach could be to notify the front end when a user is on the front end task list section, and fire an event to the backend. The backend can then just note that the front end sent a user editing `___` task list section or left task list section and update accordingly.  This would add control flow logic to the backend, instead of requiring the front end to handle locking / unlocking imperatively.

I also think that we should enforce a task list locked section on the backend, eg handle an event where a user tries to update a section that was locked by someone else. I didn't see any logic that would prevent that currently.

File: pkg/graph/resolvers/plan_task_list_section_locks.go (hunk 4 above)
Comment (OddTomBrooks) [reply pkg/graph/resolvers/plan_task_list_section_locks.go:144]: There are three main things to address here:

- Why have I made these wrapper functions which simply create and use an object?
-- In order to provide a similar retroactive structure compliant with our other resolvers
- Ok sure, but why create an object? Why not just do the same as we have been doing?
-- Mostly to facilitate the generation of a mock resolver, see pkg/graph/resolvers/mockplantasklistresolver/resolver.go
- What value does the mock resolver provide?
-- As it is, not too much. Currently we can use the mock resolver to satisfy system level testing. However, by building out a system where the server is provided with a mapping of resolvers we greatly improve the flexibility and testability of our servers. That is outside the scope of this PR, having enough side complexity as-is.

TL;DR:
Upon the development of system where GraphQL can access provided resolver instances, these convenience/wrapper functions will be obsolete and should be removed.

**Comment by OddTomBrooks** [comment]
> A lot of great stuff here! I think that this subscription system will open a lot of doors for us. I still need to review more, but I have a little bit of feedback (see the comments below). I wasn't able to test the subscription portion yet, could you provide a bit more data about what the URL you send the web socket init and subscribe methods to?
> 
> I would also really like to see an ADR and some sort of documentation for this. From our conversations about subscriptions, I know you have put a lot of excellent thought into why you went with this technology, and considerations we should have for the future. I think it is important to document it here while it is still fresh.
> 
> Can you also speak to how you envision the task list locking to go? It seems that this really is just a method for the front end to call, but we don't enforce it anywhere on the backend. I think it could be worth a bit of extra time on this portion.
> 
> Another possible approach could be to notify the front end when a user is on the front end task list section, and fire an event to the backend. The backend can then just note that the front end sent a user editing `___` task list section or left task list section and update accordingly. This would add control flow logic to the backend, instead of requiring the front end to handle locking / unlocking imperatively.
> 
> I also think that we should enforce a task list locked section on the backend, eg handle an event where a user tries to update a section that was locked by someone else. I didn't see any logic that would prevent that currently.

**URL to test subscriptions:**
ws://localhost:8085/api/graph/query

**ADR:**
Yep, agreed. I considered the ADR to be part of the Jira ticket but not PR. If we'd rather have an ADR before the PR is merged, no problem.

**Task List / Flow Control Responsibility:**
The API design on this is intended to be simple and straightforward. Except for constraining the sections to an enum rather than raw strings. The reasoning is to expect a degree of design iteration on this feature. It's far from bulletproof, but as long as an admin can go in and clear the locks it shouldn't create blockers for users through MVP.

One of the key things we'll need to properly support task section locking and unlocking in a secure way is to make it so:
- Users cannot lock more than one section at a time
- There will be an OnDisconnect event called when a user disconnects
  - Remove all user section locks
  
As for the server doing the locking on the client's behalf, I'm not sure how practical that is in a stateless environment. I suspect it is the ideal solution in a vacuum, but with everything else in context I'm not sure that our tech stack is well-positioned to make assertions about users page locality. Maybe it's possible to build a middle layer that checks page requests but that's presumptive and non-trivial.

File: pkg/graph/schema.resolvers.go (hunk 2 above)
Comment (OddTomBrooks) [reply pkg/graph/schema.resolvers.go]: Great point. This is updated for Lock and Unlock along with descriptive errors for the cases they can fail. There is also improved feedback from the admin endpoint UnlockAllTaskListSections to return the list of section locks removed.

File: pkg/graph/schema.graphql (hunk 3 above)
Comment (OddTomBrooks) [reply pkg/graph/schema.graphql]: Yep, agreed. Done.

**Comment by ClayBenson94** [comment]
Closing for now, as this was a Spike ticket and helped us understand some new angles/approaches!
